lib/spack/spack/relocate.py

```diff
@@ -479,3 +479,111 @@ def substitute_rpath(orig_rpath, topdir, new_root_path):
         new_rpath = path.replace(topdir, new_root_path)
         new_rpaths.append(new_rpath)
     return new_rpaths
+
+
+def is_relocatable(spec):
+    """Returns True if an installed spec is relocatable.
+
+    Args:
+        spec (Spec): spec to be analyzed
+
+    Returns:
+        True if the binaries of an installed spec
+        are relocatable and False otherwise.
+
+    Raises:
+        ValueError: if the spec is not installed
+    """
+    if not spec.install_status():
+        raise ValueError('spec is not installed [{0}]'.format(str(spec)))
+
+    # Explore the installation prefix of the spec
+    for root, dirs, files in os.walk(spec.prefix, topdown=True):
+        dirs[:] = [d for d in dirs if d not in ('.spack', 'man')]
+        abs_files = [os.path.join(root, f) for f in files]
+        if not all(file_is_relocatable(f) for f in abs_files if is_binary(f)):
+            # If any of the file is not relocatable, the entire
+            # package is not relocatable
+            return False
+
+    return True
+
+
+def file_is_relocatable(file):
+    """Returns True if the file passed as argument is relocatable.
+
+    Args:
+        file: absolute path of the file to be analyzed
+
+    Returns:
+        True or false
+
+    Raises:
+
+        ValueError: if the file does not exist or the path is not absolute
+    """
+    if platform.system().lower() != 'linux':
+        msg = 'function currently implemented only for linux'
+        raise NotImplementedError(msg)
+
+    if not os.path.exists(file):
+        raise ValueError('{0} does not exist'.format(file))
+
+    if not os.path.isabs(file):
+        raise ValueError('{0} is not an absolute path'.format(file))
+
+    strings = Executable('strings')
+    patchelf = Executable('patchelf')
+
+    # Remove the RPATHS from the strings in the executable
+    set_of_strings = set(strings(file, output=str).split())
+
+    m_type, m_subtype = mime_type(file)
+    if m_type == 'application' and m_subtype != 'x-archive':
+        rpaths = patchelf('--print-rpath', file, output=str).strip()
+        set_of_strings.discard(rpaths.strip())
+
+    if any(spack.store.layout.root in x for x in set_of_strings):
+        # One binary has the root folder not in the RPATH,
+        # meaning that this spec is not relocatable
+        msg = 'Found "{0}" in {1} strings'
+        tty.debug(msg.format(spack.store.layout.root, file))
+        return False
+
+    return True
+
+
+def is_binary(file):
+    """Returns true if a file is binary, False otherwise
+
+    Args:
+        file: file to be tested
+
+    Returns:
+        True or False
+    """
+    m_type, _ = mime_type(file)
+
+    msg = '[{0}] -> '.format(file)
+    if m_type == 'application':
```

Comment:
This should check the second part of mime_type for
x-archive, x-sharedlib or x-executable
or exclude files where the second part of mime_type is octet-stream

Reply:
This is what works for me

> 

-    if m_type == 'application' and m_subtype != 'x-archive':
-        rpaths = patchelf('--print-rpath', file, output=str).strip()
-        set_of_strings.discard(rpaths.strip())
+    if m_type == 'application':
+        tty.debug('{0},{1}'.format(m_type, m_subtype))
+        if m_subtype == 'x-executable' or m_subtype == 'x-sharedlib':
+            rpaths = patchelf('--print-rpath', file, output=str).strip()
+            set_of_strings.discard(rpaths.strip())

>

Reply:
@gartung is this comment meant to apply to `file_is_relocatable`? `set_of_strings` doesn't exist in this function.

Reply:
yes